File: python/camera_ui_sdk/sensor/switch.py
from __future__ import annotations

from collections.abc import Mapping
from enum import StrEnum
from typing import Any, Generic, Literal, Protocol, overload, runtime_checkable

from typing_extensions import TypedDict, TypeVar

from ..observable import Observable
from .base import Sensor, SensorCategory, SensorLike, SensorType


class SwitchProperty(StrEnum):
    """Properties for switch controls."""

    On = "on"


class SwitchControlProperties(TypedDict):
    """Property value map for switch controls."""

    on: bool


class SwitchPropertyChangeData(TypedDict):
    """Emitted on SwitchControlLike.onPropertyChanged."""

    property: str  # SwitchProperty value
    value: bool


TStorage = TypeVar("TStorage", bound=Mapping[str, Any], default=dict[str, Any])


@runtime_checkable
class SwitchControlLike(SensorLike, Protocol):
    """Read-only proxy interface for a switch control."""

    @property
    def type(self) -> SensorType:
        return SensorType.Switch

    @overload
    def getValue(self, property: Literal[SwitchProperty.On]) -> bool | None: ...
    @overload
    def getValue(self, property: str) -> object | None: ...

    @property
    def onPropertyChanged(self) -> Observable[SwitchPropertyChangeData]: ...


class SwitchControl(Sensor[SwitchControlProperties, TStorage, str], Generic[TStorage]):
    """Generic on/off switch control. Override `setOn()` / `setOff()` to drive
    hardware and call `await super().setOn()` / `await super().setOff()` after
    success to sync the SDK state. For hardware-pushed updates, call the super
    methods from your event handler — that bypasses any plugin override and only
    syncs state.
    """

    _requires_frames = False

    def __init__(self, name: str = "Switch") -> None:
        super().__init__(name)
        self._write_state({SwitchProperty.On.value: False})

    @property
    def type(self) -> SensorType:
        return SensorType.Switch

    @property
    def category(self) -> SensorCategory:
        return SensorCategory.Control

    @property
    def on(self) -> bool:
        return bool(self.props.on)

    async def setOn(self) -> None:
        """Turn the switch on. Override to drive hardware and call
        `await super().setOn()` after success to sync the SDK state.

        Example:
            ```python
            await sw.setOn()
            ```
        """
        self._write_state({SwitchProperty.On.value: True})

    async def setOff(self) -> None:
        """Turn the switch off. Override to drive hardware and call
        `await super().setOff()` after success to sync the SDK state.

        Example:
            ```python
            await sw.setOff()
            ```
        """
        self._write_state({SwitchProperty.On.value: False})

    async def updateValue(self, property: str, value: Any) -> None:
        """Routes generic property writes to semantic methods."""
        if property == SwitchProperty.On.value:
            if value:
                await self.setOn()
            else:
                await self.setOff()
        # Unknown / non-writable property — ignored.

File: node/src/sensor/switch.ts
import { Sensor, SensorType, SensorCategory } from './base.js';
import { defineSensor, SensorDomain } from './meta.js';

import type { Observable } from '../observable/index.js';
import type { PropertyChangeOf, SensorLike } from './base.js';

/**
 * Properties for switch controls
 *
 * @internal
 */
export enum SwitchProperty {
  /** Whether the switch is on */
  On = 'on',
}

/**
 * Property value map for switch controls.
 *
 * @internal
 */
export interface SwitchControlProperties {
  [SwitchProperty.On]: boolean;
}

/** Read-only proxy interface for a switch control */
export interface SwitchControlLike extends SensorLike {
  readonly type: SensorType.Switch;
  readonly onPropertyChanged: Observable<PropertyChangeOf<SwitchControlProperties>>;

  getValue(property: SwitchProperty.On): boolean | undefined;
  getValue(property: string): unknown;
}

/**
 * Generic on/off switch control. Override `setOn()` / `setOff()` to drive
 * hardware and call `await super.setOn()` / `await super.setOff()` after
 * success to sync the SDK state. For hardware-pushed updates, call the `super`
 * methods from your event handler.
 */
export class SwitchControl<TStorage extends object = Record<string, any>> extends Sensor<SwitchControlProperties, TStorage> {
  readonly type = SensorType.Switch;
  readonly category = SensorCategory.Control;

  constructor(name = 'Switch') {
    super(name);

    this._writeState({ [SwitchProperty.On]: false });
  }

  get on(): boolean {
    return this.props.on;
  }

  /**
   * Turn the switch on. Override to drive hardware and call `await super.setOn()`
   * after success to sync the SDK state.
   *
   * @example
   * ```ts
   * await sw.setOn();
   * ```
   */
  async setOn(): Promise<void> {
    this._writeState({ [SwitchProperty.On]: true });
  }

  /**
   * Turn the switch off. Override to drive hardware and call `await super.setOff()`
   * after success to sync the SDK state.
   *
   * @example
   * ```ts
   * await sw.setOff();
   * ```
   */
  async setOff(): Promise<void> {
    this._writeState({ [SwitchProperty.On]: false });
  }

  /**
   * Cross-process consumer entry point. Dispatches writable properties
   * to semantic methods so plugin overrides (hardware actions) are honored.
   * Unknown properties are ignored — only `On` is externally writable.
   *
   * @param property - Property name to write.
   *
   * @param value - New value for the property.
   *
   * @internal
   */
  async updateValue(property: string, value: unknown): Promise<void> {
    if ((property as SwitchProperty) === SwitchProperty.On) {
      if (value) await this.setOn();
      else await this.setOff();
    }
    // Unknown / non-writable property — ignored.
  }
}

/** Registry metadata for {@link SwitchControl}. */
export const switchMeta = defineSensor({
  type: SensorType.Switch,
  category: SensorCategory.Control,
  assignmentKey: 'switch',
  multiProvider: true,
  isDetectionType: false,
  properties: {
    [SwitchProperty.On]: { type: 'boolean', writable: true },
  },
  shortcutable: true,
  cascadeTrigger: { property: SwitchProperty.On, value: true, sustained: true },
  virtual: { properties: { [SwitchProperty.On]: false } },
  semantics: {
    domain: SensorDomain.Switch,
    stateProperty: SwitchProperty.On,
    commandProperty: SwitchProperty.On,
  },
});
